### PR TITLE
Create tests for simple transformer text classification

### LINKED ISSUE #29: create tests for simple_transformer_text_classification taskmodule from pytorch-ie
This requires to write individual tests for all functions in that taskmodule (@akashkumar25 please finish the list):
 - [x] the taskmodule creation
 - [x]  `_prepare`
 - [x] `_post_prepare`
 - [x] `encode_input`
 - [x] `encode_target`
 - [x] `collate`
 - [x] `unbatch_output`
 - [x] `create_annotations_from_output`

Each test should be combined with a respective fixture function that creates the actual output. Then, the test function gets the fixture as input and checks for the correct values. E.g.:
```python
@pytest.fixture(scope="module")
def taskmodule():
  return SimpleTransformerTextClassificationTaskModule(tokenizer_name_or_path="bert-base-uncased")

def test_taskmodule(taskmodule):
  assert taskmodule is not None
  # anything else to assert, if necessary

# similar for fixture "taskmodule_prepared" and its test "test_taskmodule_prepared" etc. 
```

Note: Take inspiration from the [tests for the seq2seq taskmodule](https://github.com/ChristophAlt/pytorch-ie/blob/main/tests/taskmodules/test_transformer_seq2seq.py).

### PR DESCRIPTION
**Pull request by akashkumar25**

### CODE REVIEW

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+@pytest.fixture(scope="module")
+def pre_prepared_taskmodule(config):
+    return SimpleTransformerTextClassificationTaskModule(
+        tokenizer_name_or_path="bert-base-uncased", **config
+    )
+
+
+def test_taskmodule(pre_prepared_taskmodule):
+    assert pre_prepared_taskmodule is not None
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
    assert pre_prepared_taskmodule is not None
    assert not pre_prepared_taskmodule.is_prepared
```

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    # add label
+    doc1.label.append(label1)
+    doc2.label.append(label2)
+    return [doc1, doc2]
+
+
+@pytest.fixture(scope="module")
+def taskmodule(pre_prepared_taskmodule, documents):
+    pre_prepared_taskmodule._prepare(documents)
+    pre_prepared_taskmodule._post_prepare()
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
    # this calls _prepare(documents) and then _post_prepare()
    pre_prepared_taskmodule.prepare(documents)
```

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    return pre_prepared_taskmodule
+
+
+def test_post_prepare(taskmodule):
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
def test_prepare(taskmodule):
```

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Looks very good! Just some minor remarks. Can you also rename `pre_prepared_taskmodule` to `unprepared_taskmodule`?

EDIT: Can you also integrate [the documentation](https://github.com/ArneBinder/pie-document-level/issues/29#issuecomment-1576738842) as comments to the respective fixtures / tests?  E.g.

```python
@pytest.fixture(scope="module", params=CONFIGS_DICT.keys())
def config(request):
    """
    This fixture is defined with the @pytest.fixture decorator and is used to provide a configuration dictionary (config) to other tests.
    It uses the CONFIGS_DICT dictionary to retrieve the configuration based on the parameter passed to the fixture (request.param).
    """
    return CONFIGS_DICT[request.param]
```

EDIT2: Maybe shorten / rewrite the texts a bit. For the above example this should be sth like: "This fixture yields the config dictionaries from the CONFIGS list. It is parametrized by the serialized configs to produce clean test case identifiers, e.g. `test_encode_input[max_length=8]`." Try to name the reasoning behind the code ("clean / readable test configuration") instead of mentioning whats directly derivable from the code ("is defined with the ... decorator", "It uses the CONFIGS_DICT..."), i.e. keep it concise as much as possible (if you think there does not happen sth that needs explanation, do not write a comment at all. But maybe then you can also write, why nothing else is happening there ;) E.g. "We do not test for xxx because it is already done in yyy ...").

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 4]
```diff
@@ -71,7 +88,12 @@ def documents():
 
 @pytest.fixture(scope="module")
 def taskmodule(pre_prepared_taskmodule, documents):
-    # this calls _prepare(documents) and then _post_prepare()
+    """
+    - Prepares the task module for the given documents.
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
    - Prepares the task module with the given documents, i.e. collect available label values.
```

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 5]
```diff
@@ -85,6 +107,11 @@ def test_prepare(taskmodule):
 
 @pytest.fixture(scope="module")
 def task_encoding_without_targets(taskmodule, documents):
+    """
+    - Encodes the input of a task without targets.
+    - Generates encoded inputs for a specific document without associated targets.
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
    - Generates input encodings for a specific task from a document, but without associated targets.
```

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 6]
```diff
@@ -273,7 +321,12 @@ def model_predict_output(batch):
 
 @pytest.fixture(scope="module")
 def unbatched_outputs(taskmodule):
-    # This can be created with model_predict_output fixture above
+    """
+    - Converts model outputs from batched to unbatched format.
+    - Helps in further processing of model outputs for individual task encodings.
+    - This can be created with model_predict_output fixture above.
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]: What does "This" refer to? Please improve that sentence.

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py [hunk 7]
```diff
@@ -288,6 +341,11 @@ def test_unpatch_output(unbatched_outputs):
 
 @pytest.fixture(scope="module")
 def annotations_from_output(taskmodule, task_encoding_without_targets, unbatched_outputs):
+    """
+    - Creates named annotations from model outputs for each task encoding.
+    - Converts model outputs into human-readable named annotations.
```
Comment (ArneBinder) [inline tests/unit/taskmodules/test_simple_transformer_text_classification.py]:
```suggestion
    - Converts the inputs (task_encoding_without_targets) and the respective model outputs (unbatched_outputs) into human-readable  annotations.
```

**Review by ArneBinder — CHANGES_REQUESTED** [review]
some minors

File: tests/unit/taskmodules/test_simple_transformer_text_classification.py (hunk 6 above)
Comment (ArneBinder) [reply tests/unit/taskmodules/test_simple_transformer_text_classification.py]: fixed in 5903433143d33cb8e8d3dc70b9b3d10c24b5f314

File: src/models/transformer_text_classification.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
+from typing import Any, Dict, MutableMapping, Optional, Tuple
+
+import torchmetrics
+from pytorch_ie.core import PyTorchIEModel
```
Comment (ArneBinder) [inline src/models/transformer_text_classification.py]: Is this whole file still necessary? Can you just use `TransformerTextClassificationModel` from pytorch-ie in the `model_predict_output` fixture? We should only add code that is really necessary to not clutter the codebase. But please test if the fixtures is working by using it in the `unbatched_outputs` fixture:
```
@pytest.fixture(scope="module")
def unbatched_outputs(taskmodule, model_predict_output):
   ...
```

**Review by ArneBinder — CHANGES_REQUESTED** [review]
I noticed one important thing, see below.

File: src/models/transformer_text_classification.py (hunk 8 above)
Comment (akashkumar25) [reply src/models/transformer_text_classification.py]: No its not necessary, I have made the changes and unbatched_outputs fixture is working with it. Should I remove the this model file?

Comment (ArneBinder) [reply src/models/transformer_text_classification.py]: ok, cool. yes, please remove it. And also please remove `model_predict_output` again from the `unbatched_outputs` fixture, that was just meant to test locally that `model_predict_output` works.

Comment (akashkumar25) [reply src/models/transformer_text_classification.py]: I have removed the transformer_text_classification.py file and model_predict_output from unbatched_outputs fixture.

**Comment by ArneBinder** [comment]
This is integrated upstream (see https://github.com/ChristophAlt/pytorch-ie/pull/282), we keep this here for documentary reasons.

**Pull request closed without merging**